src/ripple/app/misc/AMMHelpers.h

```diff
@@ -145,12 +149,184 @@ withinRelativeDistance(Amt const& calc, Amt const& req, Number const& dist)
+namespace {
+
+/** Smallest solution for quadratic equation.
+ * Assumes
+ * b > 0 & 4*a*c < 0 || b < 0 & 4*a*c > 0
+ * x = (-b + sqrt(b**2 - 4*a*c))/(2*a), if b > 0
+ * x = (-b - sqrt(b**2 - 4*a*c))/(2*a), if b < 0
+ */
+inline std::optional<Number>
+solveQuadraticEqSmallest(Number const& a, Number const& b, Number const& c)
+{
+    if (auto const r = b * b - 4 * a * c; r < 0)
+        return std::nullopt;
+    else if (b > 0)
+        return (-b + root2(r)) / (2 * a);
+    else
+        return (-b - root2(r)) / (2 * a);
+}
+
+/** Generate AMM offer starting with takerGets when AMM pool
+ * from the payment perspective is IOU(in)/XRP(out)
+ * Equations:
+ * Spot Price Quality after the offer is consumed:
+ *     Qsp = (O - o) / (I + i) (1)
+ *  where O is poolPays, I is poolGets, o is takerGets, i is takerPays
+ * Swap out:
+ *     i = (I*o) / (O - o)*f (2)
+ *  where f is (1 - tfee/100000), tfee is in basis points
+ * Effective price targetQuality:
+ *     Qep = o / i (3)
+ * There are two scenarios to consider
+ * A) Qsp = Qep. Substitute i in (1) with (2) and solve for o
+ *    and Qsp = targetQuality(Qt):
+ *     Qt = (O - o) / (I + (I*o) / (O - o)*f) =>
+ *     Qt = ((O - o)**2 * f) / (I*O*f - I*o*f + I*o) =>
+ *     Qt*(I*O*f - I*o*f + I*o) = (O**2 -2*O*o + o**2)*f =>
+ *     Qt*(I*O - I*o + I*o/f) = O**2 -2*O*o + o**2 =>
+ *     o**2 + o*(I*Qt*(1 - 1 / f) - 2*O) + O**2 - Qt*I*O = 0 =>
+ *     a = 1
+ *     b = I*Qt*(1 - 1 / f) - 2*O
+ *     c = O**2 - Qt*I*O
+ *     f <= 1 => b < 0
+ *     Qt < O/I => (O - I*Qt) > 0 => c > 0 => 4*a*c > 0 =>
+ *     sqrt(b**2 - 4*a*c) < b
+ *     (see solveQuadraticEqSmallest above)
+ * B) Qep = Qsp. Substitute i in (3) with (2) and solve for o
+ *    and Qep = targetQuality(Qt):
+ *     Qt = o / ((I*o) / (O - o)*f) =>
+ *     Qt = (O - o)*f / I =>
+ *     Qt*I = O*f - o*f =>
+ *     o = O - I * Qt / f
+ * A is solved first. If o is negative then Qsp can not be
+ * matched to Qt, return nullopt. If o is positive then B is solved.
+ * If o is negative then Qep can not be matched to Qt,
+ * return nullopt. Otherwise use the smallest o. If A solution is
+ * smaller, then Qep > Qsp and Qsp == Qt, otherwise Qsp > Qep
+ * and Qep = Qt. o is rounded down to increase the offer quality
+ * and takerPays is calculated with swap out eq (2).
+ */
+template <typename TIn, typename TOut>
+std::optional<TAmounts<TIn, TOut>>
+getAMMOfferStartWithTakerGets(
+    TAmounts<TIn, TOut> const& pool,
+    Quality const& targetQuality,
+    std::uint16_t const& tfee)
+{
+    assert(targetQuality.rate() != beast::zero);
+    auto const f = feeMult(tfee);
+    auto const a = 1;
+    auto const b = pool.in * (1 - 1 / f) / targetQuality.rate() - 2 * pool.out;
+    auto const c =
+        pool.out * pool.out - pool.in * pool.out / targetQuality.rate();
+
+    auto nTakerGets = solveQuadraticEqSmallest(a, b, c);
+    if (!nTakerGets || *nTakerGets <= 0)
+        return std::nullopt;
```

Comment:
Needs two more test cases, one for `nTakerGets` being `nullopt` (i.e. when the solution is less than zero, as commended in `solveQuadraticEqSmallest`) and other when `nTakerGets` is exactly 0